scipy/spatial/transform/_rotation_groups.py

```diff
@@ -0,0 +1,125 @@
+import numpy as np
+from . import Rotation
+from scipy.constants import golden as phi
+
+
+_iI = Rotation.from_quat([0.5, 0.5 / phi, 0.5 * phi, 0])
+_iO = Rotation.from_quat([0, 1 / np.sqrt(2), 1 / np.sqrt(2), 0])
+_iT = Rotation.from_quat([1, 0, 0, 0])
+_w = Rotation.from_quat([0.5, 0.5, 0.5, -0.5])
+
+
+def _generate_group(generators):
+
+    # Perform a breadth-first search to find group elements of the binary
+    # quaternion group.
+    tol = 1E-10
+    identity = Rotation.from_quat([0, 0, 0, 1])
+    queue = [identity] + generators
+    visited = np.array([identity.as_quat()])
+    while queue:
+        node = queue.pop(0)
+        mindist = np.min(np.linalg.norm(node.as_quat() - visited, axis=1))
+        if mindist > tol:
+            visited = np.concatenate((visited, [node.as_quat()]))
+            for e in generators:
+                queue.append(node * e)
+
+    # Discard half the binary quaternion group to get the unique elements of
+    # the rotation group.
+    anchor = [1, 10, 100, 1000]
+    indices = np.where(np.dot(visited, anchor) > 0)[0]
+    visited = visited[indices]
+
+    # Clean up numerical noise for cleaner printing.
+    indices = np.where(np.abs(visited) < tol)
+    visited[indices] *= 0
+
+    return Rotation.from_quat(visited)
+
+
+def _generate_cyclic_group(n):
+
+    g = np.zeros((n, 4))
+    thetas = np.linspace(0, np.pi, n + 1)[:n]
+    g[:, 2] = np.sin(thetas)
+    g[:, 3] = np.cos(thetas)
+    return Rotation.from_quat(g)
+
+
+def _generate_dihedral_group(n):
+
+    g = np.zeros((2 * n, 4))
+    thetas = np.linspace(0, np.pi, n + 1)[:n]
+    g[:n, 2] = np.sin(thetas)
+    g[:n, 3] = np.cos(thetas)
+    g[-n:, 0] = np.cos(thetas)
+    g[-n:, 1] = np.sin(thetas)
+    return Rotation.from_quat(g)
+
+
+_groups = {'C1': _generate_cyclic_group(1),
```

Comment:
Correct me if I'm wrong, but wouldn't we prefer to avoid generating this Python dictionary object when this module is imported? Perhaps the same can be said for the top-level `_iI` and so on, variable assignments.

Maybe I just have too much of a "purist library" mindset though, where organizing things in functions & classes until they are used directly is the default thought process.

Reply:
I will change the code to an entirely class-based structure.  I should have thought this through before commenting on your review.  Sorry about this!

Reply:
Thanks for the in-depth review @tylerjereddy.

I have thought about your comments on organization and reworked the code into a purely class and function based structure with no top-level variables.  I have used decorators to make to avoid generating things until they are needed.

The class achieves achieves result that I would like:
- lazy generation of attributes for parameter-free groups
- methods for parametrized groups

I am not particularly proficient at object-oriented code though, so more input here is very welcome.